binrw_derive/src/binrw/codegen/read_options/enum.rs

```diff
@@ -129,18 +129,30 @@ fn generate_unit_enum_magic(reader_var: &TokenStream, variants: &[UnitEnumField]
                 #body
             })() {
                 v @ Ok(_) => return v,
-                Err(#TEMP) => { #RESTORE_POSITION_VARIANT(#reader_var, #POS, #TEMP)?; }
+                Err(#TEMP) => {
+                    if !#RESTORE_POSITION_VARIANT(#reader_var, #POS, #TEMP)?.is_eof() {
+                        #ALL_EOF = false;
+                    }
+                }
             }
         }
     });
 
     let return_error = quote! {
-        Err(#BIN_ERROR::NoVariantMatch {
-            pos: #POS
-        })
+        if #ALL_EOF {
+            Err(#BIN_ERROR::Io(#IO_ERROR::new(
```

Comment:
There is a helper function `not_enough_bytes` that can be adapted rather than duplicating the manual construction of this error.

Reply:
Thanks, yeah, I figured we'd want to reuse that function; wasn't sure where it should go, so I just copy/pasted to move on.